From fe67aebd5216e7d3afa1d7dee2a3f82e548556f3 Mon Sep 17 00:00:00 2001
From: Tom <54514587+GAtom22@users.noreply.github.com>
Date: Sat, 28 Jun 2025 02:59:30 -0300
Subject: [PATCH] Merge commit from fork

* fix(ante): enforce fee denom

* Revert "fix(ante): enforce fee denom"

This reverts commit ecce81015ac810156407297cf17066ad011ba322.

* update fix

* add upgrade handler

* wire upgrade

* update cmt

* small refactor

* add chlog

* add chlog
---
 CHANGELOG.md                   |  6 +++++-
 app/include_upgrade_mainnet.go | 25 +++++++++++++++----------
 app/include_upgrade_testnet.go |  2 ++
 app/keepers/keepers.go         | 12 ++++++------
 app/upgrades/v2_2/upgrades.go  | 30 ++++++++++++++++++++++++++++++
 5 files changed, 58 insertions(+), 17 deletions(-)
 create mode 100644 app/upgrades/v2_2/upgrades.go

diff --git a/CHANGELOG.md b/CHANGELOG.md
index 553774426..b9c7fef0d 100644
--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -35,7 +35,11 @@ All notable changes to this project will be documented in this file.
 
 The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/)
 
-## Unreleased
+## v2.2.0
+
+### Improvements
+
+- [GHSA-56j4-446m-qrf6](https://github.com/babylonlabs-io/babylon-ghsa-56j4-446m-qrf6/pull/1) Add bank restriction for fee collector account
 
 ## v2.1.0
 
diff --git a/app/include_upgrade_mainnet.go b/app/include_upgrade_mainnet.go
index 69efed31c..1b5bad918 100644
--- a/app/include_upgrade_mainnet.go
+++ b/app/include_upgrade_mainnet.go
@@ -8,6 +8,7 @@ import (
 	"github.com/babylonlabs-io/babylon/v2/app/upgrades/v1/mainnet"
 	v1_1 "github.com/babylonlabs-io/babylon/v2/app/upgrades/v1_1"
 	v2 "github.com/babylonlabs-io/babylon/v2/app/upgrades/v2"
+	v22 "github.com/babylonlabs-io/babylon/v2/app/upgrades/v2_2"
 )
 
 var WhitelistedChannelsID = map[string]struct{}{
@@ -20,15 +21,19 @@ var WhitelistedChannelsID = map[string]struct{}{
 	"channel-6": struct{}{},
 }
 
-// init is used to include v1 upgrade for mainnet data
+// init is used to include v2.2 upgrade for mainnet data
 func init() {
-	Upgrades = []upgrades.Upgrade{v2.CreateUpgrade(false, WhitelistedChannelsID), v1_1.Upgrade, v1.CreateUpgrade(v1.UpgradeDataString{
-		BtcStakingParamsStr:       mainnet.BtcStakingParamsStr,
-		FinalityParamStr:          mainnet.FinalityParamStr,
-		IncentiveParamStr:         mainnet.IncentiveParamStr,
-		CosmWasmParamStr:          mainnet.CosmWasmParamStr,
-		NewBtcHeadersStr:          mainnet.NewBtcHeadersStr,
-		TokensDistributionStr:     mainnet.TokensDistributionStr,
-		AllowedStakingTxHashesStr: mainnet.AllowedStakingTxHashesStr,
-	}, mainnet.ParamUpgrade)}
+	Upgrades = []upgrades.Upgrade{
+		v22.Upgrade,
+		v2.CreateUpgrade(false, WhitelistedChannelsID),
+		v1_1.Upgrade,
+		v1.CreateUpgrade(v1.UpgradeDataString{
+			BtcStakingParamsStr:       mainnet.BtcStakingParamsStr,
+			FinalityParamStr:          mainnet.FinalityParamStr,
+			IncentiveParamStr:         mainnet.IncentiveParamStr,
+			CosmWasmParamStr:          mainnet.CosmWasmParamStr,
+			NewBtcHeadersStr:          mainnet.NewBtcHeadersStr,
+			TokensDistributionStr:     mainnet.TokensDistributionStr,
+			AllowedStakingTxHashesStr: mainnet.AllowedStakingTxHashesStr,
+		}, mainnet.ParamUpgrade)}
 }
diff --git a/app/include_upgrade_testnet.go b/app/include_upgrade_testnet.go
index 37e6b00b7..b517e9be6 100644
--- a/app/include_upgrade_testnet.go
+++ b/app/include_upgrade_testnet.go
@@ -11,6 +11,7 @@ import (
 	v1rc8 "github.com/babylonlabs-io/babylon/v2/app/upgrades/v1rc8/testnet"
 	v1rc9 "github.com/babylonlabs-io/babylon/v2/app/upgrades/v1rc9/testnet"
 	v2 "github.com/babylonlabs-io/babylon/v2/app/upgrades/v2"
+	v22 "github.com/babylonlabs-io/babylon/v2/app/upgrades/v2_2"
 	v2rc4 "github.com/babylonlabs-io/babylon/v2/app/upgrades/v2rc4/testnet"
 )
 
@@ -33,5 +34,6 @@ func init() {
 		v1rc5.CreateUpgrade(),
 		v1rc8.CreateUpgrade(),
 		v1rc9.CreateUpgrade(),
+		v22.Upgrade,
 	}
 }
diff --git a/app/keepers/keepers.go b/app/keepers/keepers.go
index c50ccf6dd..77df6739b 100644
--- a/app/keepers/keepers.go
+++ b/app/keepers/keepers.go
@@ -6,6 +6,7 @@ import (
 	"path/filepath"
 	"strings"
 
+	"cosmossdk.io/errors"
 	"cosmossdk.io/log"
 	storetypes "cosmossdk.io/store/types"
 	circuitkeeper "cosmossdk.io/x/circuit/keeper"
@@ -93,8 +94,7 @@ import (
 	tokenfactorytypes "github.com/strangelove-ventures/tokenfactory/x/tokenfactory/types"
 )
 
-var errBankRestrictionDistribution = fmt.Errorf("the distribution address %s can only receive bond denom %s",
-	appparams.AccDistribution.String(), appparams.DefaultBondDenom)
+var errBankRestriction = fmt.Errorf("can only receive bond denom %s", appparams.DefaultBondDenom)
 
 // Enable all default present capabilities.
 var tokenFactoryCapabilities = []string{
@@ -722,9 +722,9 @@ func initParamsKeeper(appCodec codec.BinaryCodec, legacyAmino *codec.LegacyAmino
 	return paramsKeeper
 }
 
-// bankSendRestrictionOnlyBondDenomToDistribution restricts that only the default bond denom should be allowed to send to distribution mod acc.
+// bankSendRestrictionOnlyBondDenomToDistribution restricts that only the default bond denom should be allowed to send to distribution and fee collector mod accs.
 func bankSendRestrictionOnlyBondDenomToDistribution(ctx context.Context, fromAddr, toAddr sdk.AccAddress, amt sdk.Coins) (newToAddr sdk.AccAddress, err error) {
-	if toAddr.Equals(appparams.AccDistribution) {
+	if toAddr.Equals(appparams.AccDistribution) || toAddr.Equals(appparams.AccFeeCollector) {
 		denoms := amt.Denoms()
 		switch len(denoms) {
 		case 0:
@@ -732,10 +732,10 @@ func bankSendRestrictionOnlyBondDenomToDistribution(ctx context.Context, fromAdd
 		case 1:
 			denom := denoms[0]
 			if !strings.EqualFold(denom, appparams.DefaultBondDenom) {
-				return nil, errBankRestrictionDistribution
+				return nil, errors.Wrapf(errBankRestriction, "address %s", toAddr)
 			}
 		default: // more than one length
-			return nil, errBankRestrictionDistribution
+			return nil, errors.Wrapf(errBankRestriction, "address %s", toAddr)
 		}
 	}
 
diff --git a/app/upgrades/v2_2/upgrades.go b/app/upgrades/v2_2/upgrades.go
new file mode 100644
index 000000000..7c9ed3934
--- /dev/null
+++ b/app/upgrades/v2_2/upgrades.go
@@ -0,0 +1,30 @@
+package v2_2
+
+import (
+	"context"
+
+	store "cosmossdk.io/store/types"
+	upgradetypes "cosmossdk.io/x/upgrade/types"
+	"github.com/cosmos/cosmos-sdk/types/module"
+
+	"github.com/babylonlabs-io/babylon/v2/app/keepers"
+	"github.com/babylonlabs-io/babylon/v2/app/upgrades"
+)
+
+// UpgradeName defines the on-chain upgrade name for the Babylon v2.2 upgrade
+const UpgradeName = "v2.2"
+
+var Upgrade = upgrades.Upgrade{
+	UpgradeName:          UpgradeName,
+	CreateUpgradeHandler: CreateUpgradeHandler,
+	StoreUpgrades: store.StoreUpgrades{
+		Added:   []string{},
+		Deleted: []string{},
+	},
+}
+
+func CreateUpgradeHandler(mm *module.Manager, configurator module.Configurator, keepers *keepers.AppKeepers) upgradetypes.UpgradeHandler {
+	return func(ctx context.Context, plan upgradetypes.Plan, fromVM module.VersionMap) (module.VersionMap, error) {
+		return mm.RunMigrations(ctx, configurator, fromVM)
+	}
+}
